File: implementations/dart/lib/src/s57.dart
library b57_s57;

import 'dart:math';
import 'dart:typed_data';

import 'package:blake3_dart/blake3_dart.dart';
import 'package:cryptography/cryptography.dart';

import 'b57.dart' as b57;
import 'errors.dart';
import 'h57.dart';
import 'id57.dart';

const int s57Version = 0x01;

class S57Config {
  final List<int> serverSecretKey;
  final List<int> environmentSalt;
  final int keyId;

  S57Config({
    required this.serverSecretKey,
    required this.environmentSalt,
    this.keyId = 1,
  });
}

class S57Keys {
  final List<int> b57Aes256Key;
  final List<int> h57Key;
  final List<int> id57Key;
  final int keyId;

  S57Keys({
    required this.b57Aes256Key,
    required this.h57Key,
    required this.id57Key,
    required this.keyId,
  });
}

class S57 {
  final List<int> _serverSecret;
  final List<int> _envSalt;
  final S57Keys keys;

  S57._(this._serverSecret, this._envSalt, this.keys);

  factory S57(S57Config config) {
    if (config.serverSecretKey.length < 32 || config.environmentSalt.isEmpty) {
      throw KeyInvalidError();
    }

    final serverSecret = List<int>.from(config.serverSecretKey);
    final envSalt = List<int>.from(config.environmentSalt);
    final keyId = config.keyId == 0 ? 1 : (config.keyId & 0xff);

    final seed = Uint8List.fromList([...serverSecret, ...envSalt]);
    final keys = S57Keys(
      b57Aes256Key: blake3DeriveKey(
        context: 'S57/B57_AES_256/v1',
        keyMaterial: seed,
        outputLength: 32,
      ),
      h57Key: blake3DeriveKey(
        context: 'S57/H57_KEY/v1',
        keyMaterial: seed,
        outputLength: 32,
      ),
      id57Key: blake3DeriveKey(
        context: 'S57/ID57_KEY/v1',
        keyMaterial: seed,
        outputLength: 32,
      ),
      keyId: keyId,
    );

    return S57._(serverSecret, envSalt, keys);
  }

  S57Keys resolveKeys([int? keyId]) {
    final effective = keyId == null ? keys.keyId : (keyId & 0xff);
    final seed = Uint8List.fromList([..._serverSecret, ..._envSalt]);
    return S57Keys(
      b57Aes256Key: blake3DeriveKey(
        context: 'S57/B57_AES_256/v1',
        keyMaterial: seed,
        outputLength: 32,
      ),
      h57Key: blake3DeriveKey(
        context: 'S57/H57_KEY/v1',
        keyMaterial: seed,
        outputLength: 32,
      ),
      id57Key: blake3DeriveKey(
        context: 'S57/ID57_KEY/v1',
        keyMaterial: seed,
        outputLength: 32,
      ),
      keyId: effective,
    );
  }

  String hash(List<int> data, [H57Length length = H57Length.len256]) {
    final bits = _resolveS57H57Bits(length);
    final requested = (bits + 7) ~/ 8;
    final out = blake3Keyed(
      Uint8List.fromList(keys.h57Key),
      Uint8List.fromList(data),
      requested,
    );
    _maskExcessBits(out, bits);
    return b57.encode(out);
  }

  String id(List<int> data, [ID57Length length = ID57Length.def]) {
    final effective = length == ID57Length.def ? ID57Length.len128 : length;
    final bits = _bitsByLength(effective);
    if (bits != 128 && bits != 256 && bits != 512) {
      throw InvalidLengthEnumError(length.value);
    }

    final requested = (bits + 7) ~/ 8;
    final out = blake3Keyed(
      Uint8List.fromList(keys.id57Key),
      Uint8List.fromList(data),
      requested,
    );
    _maskExcessBits(out, bits);
    return b57.encode(out);
  }

  String random() => _encodeR57128(_readEntropyBytes(16));

  String randomTime() {
    final nowSecs = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    final timeBytes = ByteData(4)..setUint32(0, nowSecs & 0xffffffff, Endian.big);
    final random = _readEntropyBytes(12);
    final mixed = blake3(Uint8List.fromList([...timeBytes.buffer.asUint8List(), ...random]), 16);
    return _encodeR57128(mixed);
  }

  String randomCounter(int counter) {
    final c = ByteData(4)..setUint32(0, counter & 0xffffffff, Endian.big);
    final random = _readEntropyBytes(12);
    final mixed = blake3(Uint8List.fromList([...c.buffer.asUint8List(), ...random]), 16);
    return _encodeR57128(mixed);
  }

  String randomSession(List<int> sessionSecret) {
    final nonce = _readEntropyBytes(16);
    final key = blake3DeriveKey(
      context: 'S57/R57_SESSION/v1',
      keyMaterial: Uint8List.fromList(sessionSecret),
      outputLength: 32,
    );
    final mixed = blake3Keyed(Uint8List.fromList(key), Uint8List.fromList(nonce), 16);
    return _encodeR57128(mixed);
  }

  String randomDevice(List<int> deviceSecret) {
    final nonce = _readEntropyBytes(16);
    final key = blake3DeriveKey(
      context: 'S57/R57_DEVICE/v1',
      keyMaterial: Uint8List.fromList(deviceSecret),
      outputLength: 32,
    );
    final mixed = blake3Keyed(Uint8List.fromList(key), Uint8List.fromList(nonce), 16);
    return _encodeR57128(mixed);
  }

  String randomDerived(List<int> masterSecret, List<int> uniqueInput) {
    final key = blake3DeriveKey(
      context: 'S57/R57_DERIVED/v1',
      keyMaterial: Uint8List.fromList(masterSecret),
      outputLength: 32,
    );
    final mixed = blake3Keyed(Uint8List.fromList(key), Uint8List.fromList(uniqueInput), 16);
    return _encodeR57128(mixed);
  }

  String randomHardened() {
    final entropy = _readEntropyBytes(16);
    return _encodeR57128(blake3(Uint8List.fromList(entropy), 16));
  }

  String randomHybrid(List<List<int>> sources) {
    if (sources.isEmpty) {
      throw InvalidInputError('random_hybrid requires at least one source');
    }
    final merged = <int>[];
    for (final source in sources) {
      merged.addAll(source);
    }
    return _encodeR57128(blake3(Uint8List.fromList(merged), 16));
  }

  Future<String> encrypt(List<int> plaintext, [List<int>? aad]) async {
    if (keys.b57Aes256Key.length != 32) {
      throw KeyInvalidError();
    }

    final nonce = _readEntropyBytes(12);
    final algorithm = AesGcm.with256bits();
    final secretKey = SecretKey(keys.b57Aes256Key);
    final box = await algorithm.encrypt(
      plaintext,
      secretKey: secretKey,
      nonce: nonce,
      aad: aad ?? const <int>[],
    );

    final envelope = <int>[
      s57Version,
      keys.keyId,
      ...nonce,
      ...box.cipherText,
      ...box.mac.bytes,
    ];
    return b57.encode(envelope);
  }

  Future<List<int>> decrypt(String encodedEnvelope, [List<int>? aad, int? expectedKeyId]) async {
    final env = b57.decode(encodedEnvelope);
    if (env.length < 30) {
      throw InvalidInputError('invalid envelope length');
    }

    final version = env[0];
    if (version != s57Version) {
      throw InvalidVersionError(version);
    }

    final keyId = env[1];
    if (expectedKeyId != null && keyId != (expectedKeyId & 0xff)) {
      throw KeyUnavailableError(keyId);
    }
    if (keyId != keys.keyId) {
      throw KeyUnavailableError(keyId);
    }

    final nonce = env.sublist(2, 14);
    final ciphertextWithTag = env.sublist(14);
    if (ciphertextWithTag.length < 16) {
      throw InvalidInputError('invalid envelope length');
    }

    final ciphertext = ciphertextWithTag.sublist(0, ciphertextWithTag.length - 16);
    final mac = Mac(ciphertextWithTag.sublist(ciphertextWithTag.length - 16));

    final algorithm = AesGcm.with256bits();
    final secretKey = SecretKey(keys.b57Aes256Key);
    try {
      return await algorithm.decrypt(
        SecretBox(ciphertext, nonce: nonce, mac: mac),
        secretKey: secretKey,
        aad: aad ?? const <int>[],
      );
    } catch (_) {
      throw AuthFailureError();
    }
  }
}

int _resolveS57H57Bits(H57Length length) {
  if (length == H57Length.hashAuto) {
    return 256;
  }
  if (length == H57Length.len128) return 128;
  if (length == H57Length.len256) return 256;
  if (length == H57Length.len512) return 512;
  throw InvalidLengthEnumError(length.value);
}

List<int> _readEntropyBytes(int length) {
  final random = Random.secure();
  return List<int>.generate(length, (_) => random.nextInt(256));
}

String _encodeR57128(List<int> raw) {
  var encoded = b57.encode(raw);
  if (encoded.length == 22) {
    return encoded;
  }

  var derived = List<int>.from(raw);
  while (encoded.length < 22) {
    derived = blake3(Uint8List.fromList([...derived, 0x57]), 16).toList();
    encoded = b57.encode(derived);
  }
  return encoded;
}

int _bitsByLength(ID57Length length) {
  const bits = {
    ID57Length.len23: 23,
    ID57Length.len29: 29,
    ID57Length.len32: 32,
    ID57Length.len47: 47,
    ID57Length.len64: 64,
    ID57Length.len70: 70,
    ID57Length.len93: 93,
    ID57Length.len128: 128,
    ID57Length.len186: 186,
    ID57Length.len256: 256,
    ID57Length.len373: 373,
    ID57Length.len512: 512,
  };
  if (!bits.containsKey(length)) {
    throw InvalidLengthEnumError(length.value);
  }
  return bits[length]!;
}

void _maskExcessBits(List<int> bytes, int bitLength) {
  if (bytes.isEmpty) {
    return;
  }
  final excess = bytes.length * 8 - bitLength;
  if (excess <= 0) {
    return;
  }
  final mask = (0xff << excess) & 0xff;
  bytes[bytes.length - 1] &= mask;
}

File: implementations/rust/src/s57.rs
use crate::b57::{decode, encode, is_canonical, is_valid};
use crate::errors::B57Error;
use crate::h57::H57Length;
use crate::id57::{id57_bits_by_length, mask_excess_bits, ID57Length};
use aes_gcm::aead::{Aead, Payload};
use aes_gcm::{Aes256Gcm, KeyInit, Nonce};
use rand::rngs::OsRng;
use rand::RngCore;

pub const S57_VERSION: u8 = 0x01;

#[derive(Debug, Clone)]
pub struct S57Config {
    pub server_secret_key: Vec<u8>,
    pub environment_salt: Vec<u8>,
    pub key_id: u8,
}

#[derive(Debug, Clone)]
pub struct S57Keys {
    pub b57_aes_256_key: [u8; 32],
    pub h57_key: [u8; 32],
    pub id57_key: [u8; 32],
    pub key_id: u8,
}

#[derive(Debug, Clone)]
pub struct S57 {
    server_secret: Vec<u8>,
    env_salt: Vec<u8>,
    pub keys: S57Keys,
}

impl S57 {
    pub fn new(mut config: S57Config) -> Result<Self, B57Error> {
        if config.server_secret_key.len() < 32 || config.environment_salt.is_empty() {
            return Err(B57Error::key_invalid());
        }
        if config.key_id == 0 {
            config.key_id = 1;
        }

        let mut s = Self {
            server_secret: config.server_secret_key,
            env_salt: config.environment_salt,
            keys: S57Keys {
                b57_aes_256_key: [0u8; 32],
                h57_key: [0u8; 32],
                id57_key: [0u8; 32],
                key_id: config.key_id,
            },
        };
        s.keys = s.resolve_keys(config.key_id)?;
        Ok(s)
    }

    pub fn resolve_keys(&self, key_id: u8) -> Result<S57Keys, B57Error> {
        let mut seed = Vec::with_capacity(self.server_secret.len() + self.env_salt.len());
        seed.extend_from_slice(&self.server_secret);
        seed.extend_from_slice(&self.env_salt);
        if seed.len() < 32 {
            return Err(B57Error::key_invalid());
        }

        let b57_aes_256_key = blake3::derive_key("S57/B57_AES_256/v1", &seed);
        let h57_key = blake3::derive_key("S57/H57_KEY/v1", &seed);
        let id57_key = blake3::derive_key("S57/ID57_KEY/v1", &seed);

        Ok(S57Keys {
            b57_aes_256_key,
            h57_key,
            id57_key,
            key_id,
        })
    }

    pub fn hash(&self, data: &[u8], length: H57Length) -> Result<String, B57Error> {
        let bits = resolve_s57_h57_bits(length)?;
        let requested = bits.div_ceil(8);
        let mut out = vec![0u8; requested];
        let mut hasher = blake3::Hasher::new_keyed(&self.keys.h57_key);
        hasher.update(data);
        hasher.finalize_xof().fill(&mut out);
        mask_excess_bits(&mut out, bits);
        Ok(encode(&out))
    }

    pub fn id(&self, data: &[u8], length: ID57Length) -> Result<String, B57Error> {
        let bits = if length == ID57Length::Default {
            128
        } else {
            id57_bits_by_length(length)?
        };
        if bits != 128 && bits != 256 && bits != 512 {
            return Err(B57Error::invalid_length_enum(length as i32));
        }

        let requested = bits.div_ceil(8);
        let mut out = vec![0u8; requested];
        let mut hasher = blake3::Hasher::new_keyed(&self.keys.id57_key);
        hasher.update(data);
        hasher.finalize_xof().fill(&mut out);
        mask_excess_bits(&mut out, bits);
        Ok(encode(&out))
    }

    pub fn random(&self) -> Result<String, B57Error> {
        Ok(encode_r57_128(&read_entropy_bytes(16)?))
    }

    pub fn random_time(&self) -> Result<String, B57Error> {
        let mut t = [0u8; 4];
        let now_secs = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map_err(|_| B57Error::invalid_input("system clock before unix epoch"))?
            .as_secs() as u32;
        t.copy_from_slice(&now_secs.to_be_bytes());

        let r = read_entropy_bytes(12)?;
        let mixed = mix_entropy_128(&[&t, &r]);
        Ok(encode_r57_128(&mixed))
    }

    pub fn random_counter(&self, counter: u32) -> Result<String, B57Error> {
        let c = counter.to_be_bytes();
        let r = read_entropy_bytes(12)?;
        let mixed = mix_entropy_128(&[&c, &r]);
        Ok(encode_r57_128(&mixed))
    }

    pub fn random_session(&self, session_secret: &[u8]) -> Result<String, B57Error> {
        let nonce = read_entropy_bytes(16)?;
        let key = blake3::derive_key("S57/R57_SESSION/v1", session_secret);
        let h = blake3::keyed_hash(&key, &nonce);
        Ok(encode_r57_128(&h.as_bytes()[..16]))
    }

    pub fn random_device(&self, device_secret: &[u8]) -> Result<String, B57Error> {
        let nonce = read_entropy_bytes(16)?;
        let key = blake3::derive_key("S57/R57_DEVICE/v1", device_secret);
        let h = blake3::keyed_hash(&key, &nonce);
        Ok(encode_r57_128(&h.as_bytes()[..16]))
    }

    pub fn random_derived(&self, master_secret: &[u8], unique_input: &[u8]) -> Result<String, B57Error> {
        let key = blake3::derive_key("S57/R57_DERIVED/v1", master_secret);
        let h = blake3::keyed_hash(&key, unique_input);
        Ok(encode_r57_128(&h.as_bytes()[..16]))
    }

    pub fn random_hardened(&self) -> Result<String, B57Error> {
        let seed = read_entropy_bytes(16)?;
        let mixed = mix_entropy_128(&[&seed]);
        Ok(encode_r57_128(&mixed))
    }

    pub fn random_hybrid(&self, sources: &[&[u8]]) -> Result<String, B57Error> {
        if sources.is_empty() {
            return Err(B57Error::invalid_input("random_hybrid requires at least one source"));
        }
        let mixed = mix_entropy_128(sources);
        Ok(encode_r57_128(&mixed))
    }

    pub fn encrypt(&self, plaintext: &[u8], aad: &[u8]) -> Result<String, B57Error> {
        let key = self.keys.b57_aes_256_key;
        let cipher = Aes256Gcm::new_from_slice(&key).map_err(|_| B57Error::key_invalid())?;

        let mut nonce = [0u8; 12];
        OsRng
            .try_fill_bytes(&mut nonce)
            .map_err(|_| B57Error::insufficient_entropy())?;

        let ciphertext_with_tag = cipher
            .encrypt(Nonce::from_slice(&nonce), Payload { msg: plaintext, aad })
            .map_err(|_| B57Error::auth_failure())?;

        let mut envelope = Vec::with_capacity(2 + nonce.len() + ciphertext_with_tag.len());
        envelope.push(S57_VERSION);
        envelope.push(self.keys.key_id);
        envelope.extend_from_slice(&nonce);
        envelope.extend_from_slice(&ciphertext_with_tag);

        Ok(encode(&envelope))
    }

    pub fn decrypt(&self, b57_string: &str, aad: &[u8], expected_key_id: Option<u8>) -> Result<Vec<u8>, B57Error> {
        let env = decode(b57_string)?;
        if env.len() < 30 {
            return Err(B57Error::invalid_input("invalid envelope length"));
        }

        let version = env[0];
        if version != S57_VERSION {
            return Err(B57Error::invalid_version(version));
        }

        let key_id = env[1];
        if let Some(expected) = expected_key_id {
            if key_id != expected {
                return Err(B57Error::key_unavailable(key_id));
            }
        }
        if key_id != self.keys.key_id {
            return Err(B57Error::key_unavailable(key_id));
        }

        let nonce = &env[2..14];
        let ciphertext_with_tag = &env[14..];

        let cipher =
            Aes256Gcm::new_from_slice(&self.keys.b57_aes_256_key).map_err(|_| B57Error::key_invalid())?;
        cipher
            .decrypt(
                Nonce::from_slice(nonce),
                Payload {
                    msg: ciphertext_with_tag,
                    aad,
                },
            )
            .map_err(|_| B57Error::auth_failure())
    }
}

pub fn s57_is_valid(s: &str) -> bool {
    is_valid(s)
}

pub fn s57_is_canonical(s: &str) -> bool {
    is_canonical(s)
}

fn resolve_s57_h57_bits(length: H57Length) -> Result<usize, B57Error> {
    let bits = match length {
        H57Length::HashAuto => 256,
        H57Length::Len128 => 128,
        H57Length::Len256 => 256,
        H57Length::Len512 => 512,
        _ => return Err(B57Error::invalid_length_enum(length as i32)),
    };
    Ok(bits)
}

fn read_entropy_bytes(length: usize) -> Result<Vec<u8>, B57Error> {
    let mut out = vec![0u8; length];
    OsRng
        .try_fill_bytes(&mut out)
        .map_err(|_| B57Error::insufficient_entropy())?;
    Ok(out)
}

fn mix_entropy_128(parts: &[&[u8]]) -> Vec<u8> {
    let mut hasher = blake3::Hasher::new();
    for part in parts {
        hasher.update(part);
    }
    let mut out = [0u8; 16];
    hasher.finalize_xof().fill(&mut out);
    out.to_vec()
}

fn encode_r57_128(raw: &[u8]) -> String {
    let mut encoded = encode(raw);
    if encoded.len() == 22 {
        return encoded;
    }

    let mut derived = raw.to_vec();
    while encoded.len() < 22 {
        derived = mix_entropy_128(&[derived.as_slice(), &[0x57]]);
        encoded = encode(&derived);
    }
    encoded
}
